### ISSUE TITLE
Support cyclic references in multi-output derivations

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nix#494]
Currently, attempting it results in an unhelpful error as described in #481. That ticket also has some discussion about why cycles could be useful.

**Comment by roberth** [comment]
This would make errors like https://github.com/srid/haskell-flake/issues/167 a non-issue, allowing for defaults that improve the closure size of all sorts of generated packages.

Enforcing the non-existence of certain output dependencies is still useful, and this can already be done using the existing [`outputChecks`](https://nixos.org/manual/nix/stable/language/advanced-attributes.html#adv-attr-outputChecks).

I do think that this is a significantly breaking change. Such derivations can be built by a newer Nix, but older Nix versions won't be able to build the derivations that make use of cyclic references. Presumably substitution also needs a change, as the cyclic store objects have to become valid atomically after their contents have all been added successfully without them being valid store objects at first.

**Comment by roberth** [comment]
Reference-following algorithms need to be checked and tested, including build scheduling and closure copying.
Algorithms that compute a closure set first and then ignore references should require no adaptation whatsoever.

Copying probably doesn't quite work yet, because each store path is added individual under the condition that its references exist. We'd either need to check these constraints in one go when receiving the existing multi-add daemon operation, or, what I think I would prefer, is to have a flag in the store db that marks it as weakly valid, to indicate that the references have not been realised yet. When the references are realised, it can be upgraded to a normal valid path. This will also benefit copy operations without cycles because it removes an order dependency, improving concurrency and utilization during the copy process.

Usages of toposort should also be evaluated, of which we have about 14 in the C++ part, and a couple in the Perl bindings (for Hydra)

**Comment by Ericson2314** [comment]
I don't think we should do this, because I think we should Merkelize closures to protect against frankenbuilds.

(Actually, with content-addressing, this is already the case. I don't want to regress from that.)

**Comment by arianvp** [comment]
In that case maybe we should close this?

**Comment by Ericson2314** [comment]
I am not the whole Nix team, but I would advocate closing this as "won't do".

**Comment by roberth** [comment]
> because I think we should Merkelize closures

Is modulo-hashing not a feasible strategy?

> protect against frankenbuilds

Have we defined that term anywhere? I think I know what this refers to, but I'd really like to not guess the definition.

How does keeping a better record of inputs inhibit us from having cycles?


> In that case maybe we should close this?

I'm not convinced.

I do think we should prioritize correctness over new features like this, e.g. solve the "frankenbuild" problem before supporting cyclic references, but based on how we've been using the issue tracker, we only close feature requests if they conflict with existing requirements.
Concretely, we should close this issue if @Ericson2314 can show a clearly unavoidable conflict between solving the frankenbuild problem and cycle support.
If we're going to decide based on vibes, we may as well not track feature requests at all (needlessly disruptive; not a serious suggestion).

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
